cmd/src/search.go

```diff
@@ -299,6 +334,115 @@ func searchHighlightPreview(preview interface{}, start, end string) string {
+	// adds).
+	uniqueStartOfMatchToken := "pXRdMhZbgnPL355429nsO4qFgX86LfXTSmqH4Nr3#*(@)!*#()@!APPJB8ZRutvZ5fdL01273i6OdzLDm0UMC9372891skfJTl2c52yR1v"
+	uniqueEndOfMatchToken := "v1Ry25c2lTJfks1982739CMU0mDLzdO6i37210Ldf5ZvtuRZ8BJPPA!@)(#*!)@(*#3rN4HqmSTXfL68XgFq4Osn924553LPngbZhMdRXp"
+	diff = searchHighlightPreview(diffPreview, uniqueStartOfMatchToken, uniqueEndOfMatchToken)
+
+	// Now highlight our diff with colordiff.
+	var buf bytes.Buffer
+	cmd := exec.Command(path)
+	cmd.Stdin = strings.NewReader(diff)
+	cmd.Stdout = &buf
+	if err := cmd.Run(); err != nil {
+		fmt.Println("warning: colordiff failed to colorize diff:", err)
+		return diff
+	}
+	colorized := buf.String()
+	var final []string
+	for _, line := range strings.Split(colorized, "\n") {
+		// fmt.Println("LINE", line)
+		// Find where the start-of-match token is in the line.
+		somToken := strings.Index(line, uniqueStartOfMatchToken)
+
+		// Find which ANSI codes are to the left of our start-of-match token.
+		indices := ansiRegexp.FindAllStringIndex(line, -1)
+		matches := ansiRegexp.FindAllString(line, -1)
+		var left []string
+		for k, index := range indices {
+			if index[0] < somToken && index[1] < somToken {
+				left = append(left, matches[k])
+			}
+		}
+
+		// Replace our start-of-match token with the color we wish.
+		line = strings.Replace(line, uniqueStartOfMatchToken, ansiColors["search-match"], -1)
+
+		// Replace our end-of-match token with the color terminator,
+		// and start all colors that were previously started to the left.
+		line = strings.Replace(line, uniqueEndOfMatchToken, ansiColors["nc"]+strings.Join(left, ""), -1)
+
+		final = append(final, line)
+	}
+	return strings.Join(final, "\n")
+}
+
+func renderResult(searchResult map[string]interface{}) string {
+	searchResultBody := searchResult["body"].(map[string]interface{})
+	text := searchResultBody["text"].(string)
+	plainText := markdownToPlainText(text)
+	highlights := searchResult["highlights"]
+	isDiff := strings.HasPrefix(text, "```diff") && strings.HasSuffix(text, "```")
+	if _, ok := highlights.([]interface{}); ok {
+		if isDiff {
+			// We special case diffs because we want to display them with color.
+			return searchHighlightDiffPreview(map[string]interface{}{"value": plainText, "highlights": highlights.([]interface{})})
+		}
+		return searchHighlightPreview(map[string]interface{}{"value": plainText, "highlights": highlights.([]interface{})}, "", "")
+	}
+	return text
+}
+
+func markdownToPlainText(input string) string {
+	renderedHTML := blackfriday.Run([]byte(input))
+	text, err := html2text.FromString(string(renderedHTML), html2text.Options{OmitLinks: true})
+	text = html.UnescapeString(text)
+	if err != nil {
+		return input
+	}
+	return text
+}
+
+// Checks the Sourcegraph instance's build version date to determine if the new search result interface exists.
+func buildVersionHasNewSearchInterface(buildVersion interface{}) bool {
+	dateRegex, err := regexp.Compile("(\\w{4}-\\w{2}-\\w{2})")
```

Comment:
Declare this regex as a package-level variable and use `regexp.MustCompile`. This is better because this function gets called _a lot_ and compiling regex isn't free